sharktank/sharktank/layers/paged_llama_attention_block.py

```diff
@@ -108,9 +137,95 @@ def __init__(
                 RMSNormLayer(theta("attn_output_norm"), epsilon=rms_epsilon),
             )
 
+    def pre_process_attention(
+        self,
+        x: torch.Tensor,
+        start_index: int,
+        embedding: RotaryEmbeddingLayer,
+        embedding_batch_mask: torch.Tensor,
+    ):
+
+        if self.attn_type == "mla":
+            if self.wq is not None:
+                q = self.wq(x).unflatten(2, (self.head_count, -1))
+            else:
+                q = self.wq_b(self.q_norm(self.wq_a(x)))
+                if not isinstance(q, ReplicatedTensor):
+                    q = ops.replicate(q, count=self.shard_count)
+                q = q.unflatten(2, (self.head_count, -1))
+
+            qk_nope_head_dim = q.shape[-1] - self.rope_dimension_count
+            q_nope = q[:, :, :, :qk_nope_head_dim]
+            q_rope = q[:, :, :, qk_nope_head_dim:]
+
+            kv = self.wkv_a(x)
+            kv_nope_size = kv.shape[-1] - self.rope_dimension_count
+            kv_nope = kv[:, :, :kv_nope_size]
+            k_rope = kv[:, :, kv_nope_size:]
+
+            if start_index is not None:
+                q_rope = embedding(xt=q_rope, start_index=start_index)
+                k_rope = embedding(xt=k_rope.unsqueeze(2), start_index=start_index)
+            else:
+                q_rope = embedding.apply_batched_mask(
+                    xt=q_rope, mask=embedding_batch_mask
+                )
+                k_rope = embedding.apply_batched_mask(
+                    xt=k_rope.unsqueeze(2), mask=embedding_batch_mask
+                )
+
+            xq = ops.cat((q_nope, q_rope), dim=-1)
+
+            ##TODO: Restructure this to apply the wkv_b post attention instead of here
+            kv_norm = self.kv_norm(kv_nope)
+            wkv_b = self.wkv_b(kv_norm)
+            if not isinstance(wkv_b, ReplicatedTensor):
```

Comment:
```suggestion
            if isinstance(wkv_b, SplitPrimitiveTensor):
```

Allows unsharded models to be exported.